File: engine/scripts/lua/tests/test_map_postprocessor.lua
local assert = require("assert")

local called = false

local function test_postprocessor_invoked()
    called = false
    world:clear_map_postprocessors()
    world:register_map_postprocessor(function(w)
        called = true
        assert.equals(w:get_map_cell_count(), 1)
    end)
    world:apply_generated_map({ topology = "square", cells = { { x = 0, y = 0, z = 0 } } })
    assert.is_true(called)
end

local function test_postprocessor_error_blocks()
    world:clear_map_postprocessors()
    world:register_map_postprocessor(function(w)
        error("fail on purpose")
    end)
    local ok, err = pcall(function()
        world:apply_generated_map({ topology = "square", cells = { { x = 0, y = 0, z = 0 } } })
    end)
    assert.is_false(ok)
    assert.is_true(tostring(err):find("fail on purpose"))
end

local function test_clear_postprocessors()
    local called2 = false
    world:register_map_postprocessor(function(w)
        called2 = true
    end)
    world:clear_map_postprocessors()
    world:apply_generated_map({ topology = "square", cells = { { x = 0, y = 0, z = 0 } } })
    assert.is_false(called2)
end

-- === Map Validator Tests ===

local function test_validator_blocks_invalid_map()
    world:clear_map_validators()
    world:register_map_validator(function(map)
        -- Always fail
        return false
    end)
    local ok, err = pcall(function()
        world:apply_generated_map({ topology = "square", cells = { { x = 0, y = 0, z = 0 } } })
    end)
    assert.is_false(ok)
    assert.is_true(tostring(err):find("Map validator failed"))
end

local function test_validator_accepts_valid_map()
    world:clear_map_validators()
    local called = false
    world:register_map_validator(function(map)
        called = true
        return true
    end)
    world:apply_generated_map({ topology = "square", cells = { { x = 0, y = 0, z = 0 } } })
    assert.is_true(called)
end

local function test_clear_validators()
    local called = false
    world:register_map_validator(function(map)
        called = true
        return true
    end)
    world:clear_map_validators()
    world:apply_generated_map({ topology = "square", cells = { { x = 0, y = 0, z = 0 } } })
    assert.is_false(called)
end

return {
    test_postprocessor_invoked = test_postprocessor_invoked,
    test_postprocessor_error_blocks = test_postprocessor_error_blocks,
    test_clear_postprocessors = test_clear_postprocessors,
    test_validator_blocks_invalid_map = test_validator_blocks_invalid_map,
    test_validator_accepts_valid_map = test_validator_accepts_valid_map,
    test_clear_validators = test_clear_validators,
}

File: engine_py/tests/test_map_postprocessor.py
import pytest

def test_register_and_invoke_map_postprocessor(make_world):
    world = make_world()
    called = {}

    def validator(world_obj):
        called["ran"] = True
        assert world_obj.get_map_cell_count() == 0 or isinstance(
            world_obj.get_map_cell_count(), int
        )
        return True

    world.register_map_postprocessor(validator)
    map_data = {"topology": "square", "cells": [{"x": 0, "y": 0, "z": 0}]}
    world.apply_generated_map(map_data)
    assert called.get("ran")

def test_postprocessor_can_block_apply(make_world):
    world = make_world()

    def always_fail(world_obj):
        raise Exception("fail on purpose")

    world.register_map_postprocessor(always_fail)
    map_data = {"topology": "square", "cells": [{"x": 0, "y": 0, "z": 0}]}
    with pytest.raises(Exception) as excinfo:
        world.apply_generated_map(map_data)
    assert "fail on purpose" in str(excinfo.value)

def test_clear_map_postprocessors(make_world):
    world = make_world()
    called = {}

    def validator(world_obj):
        called["ran"] = True
        return True

    world.register_map_postprocessor(validator)
    world.clear_map_postprocessors()
    map_data = {"topology": "square", "cells": [{"x": 0, "y": 0, "z": 0}]}
    world.apply_generated_map(map_data)
    assert not called.get("ran", False)

# === Map Validator Tests ===

def test_validator_blocks_invalid_map(make_world):
    world = make_world()
    world.clear_map_validators()

    def always_fail(map_obj):
        return False

    world.register_map_validator(always_fail)
    map_data = {"topology": "square", "cells": [{"x": 0, "y": 0, "z": 0}]}
    with pytest.raises(Exception) as excinfo:
        world.apply_generated_map(map_data)
    assert "Map validator failed" in str(excinfo.value)

def test_validator_accepts_valid_map(make_world):
    world = make_world()
    world.clear_map_validators()
    called = {}

    def always_pass(map_obj):
        called["ran"] = True
        return True

    world.register_map_validator(always_pass)
    map_data = {"topology": "square", "cells": [{"x": 0, "y": 0, "z": 0}]}
    world.apply_generated_map(map_data)
    assert called.get("ran")

def test_clear_validators(make_world):
    world = make_world()
    called = {}

    def validator(map_obj):
        called["ran"] = True
        return True

    world.register_map_validator(validator)
    world.clear_map_validators()
    map_data = {"topology": "square", "cells": [{"x": 0, "y": 0, "z": 0}]}
    world.apply_generated_map(map_data)
    assert not called.get("ran", False)
